### PR TITLE
Add dedicated SszBitlist and its schema

### PR DESCRIPTION
**Pull request by Nashatyrev**
## PR Description

- Add `SszBitlist` and `SszBitlistSchema`
- Replace legacy `Bitlist` usages with `SszBitlist`
- Though `Bitlist` is still internally used it is now not exposed (package private)
- Add minimal unit tests set 
- Remove obsolete `SszComplexSchemas.SszBitListSchema`
- Make lazy `TreeNode` calculation for `SszBitlist` to avoid overhead when using `SszBitlist` as just a bitlist structure
- Bonus refactor: extract `SszContainerSchema` interface

## TODO 
- [x] check benchmarks
- [x] add more `SszBitlist` and `SszBitlistSchema` unit tests

## Benchmarks
(`BitlistBenchmark` in its old form is no more relevant, so was changed and results can't be compared)
| Benchmark                                                                        | Score Before | Error       | Units  | Score After | Error       |
|----------------------------------------------------------------------------------|--------------|-------------|--------|-------------|-------------|
| t.p.t.benchmarks.SSZBenchmark.BeaconStateSerialization                           | 779.608      | 63.932      | ops/s  | 766.650     | 66.504      |
| t.p.t.benchmarks.Sha256Benchmark.sha256of33byteArray                             | 1420784.021  | 28402.691   | ops/s  | 1389075.321 | 32640.610   |
| t.p.t.benchmarks.Sha256Benchmark.sha256of33bytes                                 | 1389238.515  | 26733.946   | ops/s  | 1381940.655 | 38107.118   |
| t.p.t.benchmarks.SpecBenchmarks.integerSquareRoot                                | 1809285.927  | 90172.211   | ops/s  | 1759424.409 | 78305.049   |
| t.p.t.benchmarks.VoteTrackerSerialize.genericDeserialization                     | 1757.969     | 416.563     | ops/ms | 1373.555    | 413.851     |
| t.p.t.benchmarks.VoteTrackerSerialize.genericSerialization                       | 4326.068     | 474.103     | ops/ms | 4180.761    | 437.056     |
| t.p.t.benchmarks.VoteTrackerSerialize.voteTrackerDeserialization                 | 1261.675     | 470.908     | ops/ms | 1089.923    | 473.447     |
| t.p.t.benchmarks.VoteTrackerSerialize.voteTrackerSerialization                   | 3157.504     | 497.497     | ops/ms | 3029.200    | 448.821     |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.iterateBalances               | 2728.907     | 39.118      | ops/s  | 3051.709    | 60.499      |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.iterateValidators             | 1971.527     | 164.796     | ops/s  | 2166.284    | 125.289     |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.iterateValidatorsWithMethods  | 70.032       | 2.590       | ops/s  | 76.144      | 1.896       |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.updateBalancesAndHash         | 40.928       | 0.841       | ops/s  | 38.795      | 0.769       |
| ~~t.p.t.benchmarks.util.backing.BitlistBenchmark.fromCachedListView~~                | 565510.637   | 14714.153   | ops/s  | 79688.884   | 3447.965    |
| ~~t.p.t.benchmarks.util.backing.BitlistBenchmark.fromNewListView~~                   | 536193.093   | 27328.012   | ops/s  | 23289.222   | 4359.102    |
| t.p.t.benchmarks.util.backing.ListBenchmark.append                               | 6067469.510  | 2982313.469 | ops/s  | 6008475.831 | 2571404.811 |
| t.p.t.benchmarks.util.backing.ListBenchmark.createDefaultUIntList                | 2591106.098  | 560768.582  | ops/s  | 3074719.821 | 709128.778  |
| t.p.t.benchmarks.util.backing.ListBenchmark.incrementalHash                      | 8916.794     | 1828.657    | ops/s  | 10753.118   | 1866.507    |
| t.p.t.benchmarks.BeaconBlockBenchmark.hashFullBlocks                             | 0.006        | 0.001       | us/op  | 0.008       | 0.001       |
| t.p.t.benchmarks.BeaconBlockBenchmark.hashSparseBlocks                           | 0.007        | 0.001       | us/op  | 0.008       | 0.001       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchCreate                         | 9.586        | 0.136       | us/op  | 12.707      | 0.425       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchCreateAndIterate               | 10.587       | 1.121       | us/op  | 12.503      | 0.434       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchCreateAndSerialize             | 15.711       | 1.883       | us/op  | 18.814      | 0.872       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchDeserialize                    | 4.075        | 0.329       | us/op  | 3.982       | 0.183       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchDeserializeAndIterate          | 10.664       | 4.667       | us/op  | 9.006       | 0.921       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchDeserializeAndSerialize        | 10.791       | 2.342       | us/op  | 9.381       | 0.684       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchIterate                        | 0.095        | 0.031       | us/op  | 0.088       | 0.004       |
| t.p.t.benchmarks.ssz.SszAttestationBenchmark.benchSerialize                      | 3.976        | 0.214       | us/op  | 4.111       | 0.398       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchCreate                     | 22.376       | 1.742       | us/op  | 21.288      | 0.890       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchCreateAndIterate           | 26.678       | 4.459       | us/op  | 24.468      | 0.865       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchCreateAndSerialize         | 78.213       | 9.073       | us/op  | 75.895      | 2.549       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchDeserialize                | 50.551       | 9.848       | us/op  | 46.920      | 3.074       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchDeserializeAndIterate      | 135.733      | 97.520      | us/op  | 87.453      | 5.286       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchDeserializeAndSerialize    | 115.376      | 14.644      | us/op  | 102.709     | 4.244       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchIterate                    | 1.757        | 0.259       | us/op  | 1.283       | 0.071       |
| t.p.t.benchmarks.ssz.SszBeaconBlockBodyBenchmark.benchSerialize                  | 52.842       | 4.186       | us/op  | 50.365      | 2.299       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchCreate                          | 0.359        | 0.022       | us/op  | 0.303       | 0.021       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchCreateAndIterate                | 0.346        | 0.025       | us/op  | 0.368       | 0.018       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchCreateAndSerialize              | 0.441        | 0.015       | us/op  | 0.462       | 0.022       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchDeserialize                     | 0.269        | 0.010       | us/op  | 0.258       | 0.013       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchDeserializeAndIterate           | 0.374        | 0.016       | us/op  | 0.362       | 0.015       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchDeserializeAndSerialize         | 0.411        | 0.029       | us/op  | 0.396       | 0.025       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchIterate                         | 0.013        | 0.001       | us/op  | 0.013       | 0.001       |
| t.p.t.benchmarks.ssz.SszCheckpointBenchmark.benchSerialize                       | 0.141        | 0.015       | us/op  | 0.131       | 0.014       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchCreate                  | 10.071       | 0.578       | us/op  | 12.147      | 1.352       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchCreateAndIterate        | 13.897       | 1.723       | us/op  | 11.952      | 0.533       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchCreateAndSerialize      | 15.802       | 1.441       | us/op  | 17.067      | 1.337       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchDeserialize             | 4.812        | 2.013       | us/op  | 3.749       | 0.346       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchDeserializeAndIterate   | 8.084        | 1.105       | us/op  | 7.119       | 0.434       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchDeserializeAndSerialize | 9.473        | 1.602       | us/op  | 8.391       | 0.284       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchIterate                 | 1.111        | 0.051       | us/op  | 0.094       | 0.004       |
| t.p.t.benchmarks.ssz.SszPendingAttestationBenchmark.benchSerialize               | 4.245        | 0.674       | us/op  | 3.661       | 0.240       |
| t.p.t.benchmarks.TransitionBenchmark.Block.importBlock                           | 0.062        | 0.004       | s/op   | 0.050       | 0.003       |
| t.p.t.benchmarks.TransitionBenchmark.Epoch.importBlock                           | 0.185        | 0.014       | s/op   | 0.178       | 0.011       |

## Fixed Issue(s)

Part of #3540 

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Comment by Nashatyrev** [comment]
Hm, having performance degradation on one benchmark: 
| Benchmark                                                                       | Old Score | Error   | Units | New Score | Error  |
|---------------------------------------------------------------------------------|-----------|---------|-------|-----------|--------|
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.iterateBalances              | 2728.907  | 39.118  | ops/s | 373.554   | 20.224 |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.iterateValidators            | 1971.527  | 164.796 | ops/s | 317.325   | 14.654 |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.iterateValidatorsWithMethods | 70.032    | 2.59    | ops/s | 54.273    | 1.721  |
| t.p.t.benchmarks.util.backing.BeaconStateBenchmark.updateBalancesAndHash        | 40.928    | 0.841   | ops/s | 40.282    | 0.437  |

Strange cause it seems to be irrelevant to bitlist. Probably some regression

**Comment by Nashatyrev** [comment]
Turned out the memorizing supplier was too expensive 🤔 
```java
  private final IntSupplier sizeCache = IntSupplier.memorizing(this::sizeImpl);
```
Replacin with regular `int` and usual condition resolved the issue: 141be7f

File: data/serializer/src/test/java/tech/pegasys/teku/api/schema/SignedBeaconBlockTest.java [hunk 1]
```diff
@@ -27,8 +28,7 @@ public void shouldConvertSchemaToInternalCorrectly() {
     final tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock internalBlock =
         dataStructureUtil.randomSignedBeaconBlock(1);
     final SignedBeaconBlock apiBlock = new SignedBeaconBlock(internalBlock);
-    assertThat(apiBlock.asInternalSignedBeaconBlock())
-        .usingRecursiveComparison()
-        .isEqualTo(internalBlock);
+    assertThat(SszTestUtils.equalsByGetters(apiBlock.asInternalSignedBeaconBlock(), internalBlock))
```
Comment (mbaxter) [inline data/serializer/src/test/java/tech/pegasys/teku/api/schema/SignedBeaconBlockTest.java]: (optional) We could also add a new assert util similar to `SafeFutureAssert` to hold ssz-related assertion helpers

**Review by mbaxter — APPROVED** [review]
LGTM

File: data/serializer/src/test/java/tech/pegasys/teku/api/schema/SignedBeaconBlockTest.java (hunk 1 above)
Comment (Nashatyrev) [reply data/serializer/src/test/java/tech/pegasys/teku/api/schema/SignedBeaconBlockTest.java]: Yeah, nice idea: 2f30ead 👍

**Pull request merged**